src/bifrost/core.clj

```diff
@@ -43,24 +38,47 @@
                                (:path-params request)
                                (:bifrost-params request))))
 
+(defn ctx->bifrost-request [ctx]
+  (let [request (:request ctx)]
+    (assoc request :bifrost-params (params-map request))))
+
+(def interceptor-xf
+  (map (fn [[response-ch ctx]]
+         [response-ch (ctx->bifrost-request ctx)])))
+
+(defn api-response->ctx
+  [api-response]
+  (let [status (response->http-status api-response)]
+    {:response (-> api-response
+                   (dissoc :status)
+                   ring-resp/response
+                   (ring-resp/status status))}))
+
+(def api-response-xf (map api-response->ctx))
+
 (defn async-interceptor
-  [channel]
-  (interceptor
-   {:enter
-    (fn [ctx]
-      (let [response-channel (async/chan 1)
-            request (:request ctx)]
-        (async/>!! channel [response-channel (params-map request)])
-        (async/go
-          (let [response (async/alt! (async/timeout *response-timeout*) timeout-response
-                                     response-channel ([r] r))
-                status (response->http-status response)]
-            (assoc ctx :response
-                   (-> response
-                       (dissoc :status)
-                       ring-resp/response
-                       (ring-resp/status status)))))))}))
+  ([channel]
+   (async-interceptor channel (gensym)))
+  ([channel response-channel-key]
+   (async-interceptor channel response-channel-key (map identity)))
+  ([channel response-channel-key response-channel-xf]
+   (let [response-channel (async/chan 1 response-channel-xf)]
+     (interceptor/interceptor
+      {:enter
+       (fn [ctx]
+         (async/>!! channel [response-channel ctx])
+         (assoc-in ctx [:response-channels response-channel-key] response-channel))
+       :leave
+       (fn [ctx]
+         (if-let [response (async/alt!!
+                             (async/timeout *response-timeout*) {:response {:status 500 :body "Timeout"}}
+                             response-channel ([r] r))]
+           (merge ctx response)
+           ctx))}))))
 
-(extend-protocol IntoInterceptor
-  clojure.core.async.impl.channels.ManyToManyChannel
-  (-interceptor [ch] (async-interceptor ch)))
```

Comment:
Since the `interceptor` macro just takes the channel as its single arg, could we not use it here in this protocol in place of `async-interceptor`?